botocore/http_session.py

```diff
@@ -0,0 +1,216 @@
+            )
+
+        return self._proxy_managers[proxy_url]
+
+    def _path_url(self, url):
+        parsed_url = urlparse(url)
+        path = parsed_url.path
+        if not path:
+            path = '/'
+        if parsed_url.query:
+            path = path + '?' + parsed_url.query
+        return path
+
+    def _setup_ssl_cert(self, conn, url, verify):
+        if url.lower().startswith('https') and verify:
+            conn.cert_reqs = 'CERT_REQUIRED'
+            conn.ca_certs = get_cert_path(verify)
+        else:
+            conn.cert_reqs = 'CERT_NONE'
+            conn.ca_certs = None
+
+    def send(self, request):
+        try:
+            request_target = self._path_url(request.url)
+
+            proxy_url = self._proxy_config.proxy_url_for(request.url)
+            if proxy_url:
+                manager = self._get_proxy_manager(proxy_url)
+            else:
+                manager = self._manager
+
+            if proxy_url and request.url.startswith('http:'):
+                # If an http request is being proxied use the full url
+                request_target = request.url
+
+            conn = manager.connection_from_url(request.url)
+            self._setup_ssl_cert(conn, request.url, self._verify)
+            urllib_response = conn.urlopen(
+                method=request.method,
+                url=request_target,
+                body=request.body,
+                headers=request.headers,
+                retries=False,
+                assert_same_host=False,
+                preload_content=False,
+                decode_content=False,
+            )
+
+            http_response = botocore.awsrequest.AWSResponse(
+                request.url,
+                urllib_response.status,
+                urllib_response.headers,
+                urllib_response,
+            )
+
+            if not request.stream_output:
+                # Cause the raw stream to be exhausted immediately. We do it
+                # this way instead of using preload_content because
+                # preload_content will never buffer chunked responses
+                http_response.content
+
+            return http_response
+        except URLLib3SSLError as e:
+            raise SSLError(endpoint_url=request.url, error=e)
+        except (NewConnectionError, socket.gaierror) as e:
+            raise EndpointConnectionError(endpoint_url=request.url, error=e)
+        except ProxyError as e:
+            raise ProxyConnectionError(proxy_url=proxy_url, error=e)
+        except URLLib3ConnectTimeoutError as e:
+            raise ConnectTimeoutError(endpoint_url=request.url, error=e)
+        except URLLib3ReadTimeoutError as e:
+            raise ReadTimeoutError(endpoint_url=request.url, error=e)
+        except ProtocolError as e:
+            raise ConnectionClosedError(
+                error=e,
+                request=request,
+                endpoint_url=request.url
+            )
+        except Exception as e:
```

Comment:
Are we sure we want to catch `Exception` here?  Is is possible to restrict this to a list of exceptions we know about?  This will consider even trivial bugs/typos as an `HTTPClientError`.

Reply:
I think so, it ensures we never directly leak any exceptions from our dependency or a mistake in our implementation. What's your concern with an exception from this layer being wrapped as an `HTTPClientError`?

Reply:
Primarily that actual bugs that raise things like `TypeError`/`ValueError` etc will be misrepresented as a `HTTPClientError`.  Ideally we'd want these things to fail fast and propagate all the way up the stack vs. possibly being retried and surfaced as a (somewhat confusing) `HTTPClientError`.

Note a huge deal, but I'd prefer to avoid it if possible.